zetta_utils/layer/volumetric/tensorstore/backend.py

```diff
@@ -0,0 +1,277 @@
+# pylint: disable=missing-docstring
+from __future__ import annotations
+
+import ast
+from copy import deepcopy
+from typing import Any, Dict, Literal, Optional, Union
+
+import attrs
+import cachetools
+import numpy as np
+import tensorstore
+import torch
+from cachetools.keys import hashkey
+
+from zetta_utils import tensor_ops
+from zetta_utils.geometry import Vec3D
+
+from .. import VolumetricBackend, VolumetricIndex
+from ..cloudvol import (
+    InfoExistsModes,
+    PrecomputedInfoSpec,
+    _get_info,
+    _info_cache,
+    _info_hash_key,
+)
+
+_ts_cache: cachetools.LRUCache = cachetools.LRUCache(maxsize=500)
+_ts_hash_key = hashkey
+
+# str for hashability
+@cachetools.cached(_ts_cache, key=_ts_hash_key)
+def _get_ts_at_resolution(path, resolution: Optional[str] = None) -> tensorstore.TensorStore:
+    spec: Dict[str, Any] = {
+        "driver": "neuroglancer_precomputed",
+        "kvstore": path,
+    }
+    if resolution is not None:
+        spec["scale_metadata"] = {"resolution": ast.literal_eval(resolution)}
+    result = tensorstore.open(spec).result()
+    return result
+
+
+@attrs.mutable
+class TSBackend(VolumetricBackend):  # pylint: disable=too-few-public-methods
+    """
+    Backend for peforming IO on Neuroglancer datasts using TensorStore library.
+    Read data will be a ``torch.Tensor`` in ``BCXYZ`` dimension order.
+    Write data is expected to be a ``torch.Tensor`` or ``np.ndarray`` in ``BCXYZ``
+    dimension order.
+    :param path: Precomputed path.
+    :param info_spec: Specification for the info file for the layer. If None, the
+        info is assumed to exist.
+    :param on_info_exists: Behavior mode for when both `info_spec` is given and
+        the layer info already exists.
+
+    """
+
+    path: str
+    info_spec: Optional[PrecomputedInfoSpec] = None
+    on_info_exists: InfoExistsModes = "expect_same"
+
+    def __attrs_post_init__(self):
+        if self.info_spec is not None:
+            new_info = self.info_spec.make_info()
+            if new_info is not None:
+                try:
+                    existing_info = _get_info(self.path)
+                except FileNotFoundError:
+                    existing_info = None
+
+                if (
+                    existing_info is not None
+                    and self.on_info_exists == "expect_same"
+                    and new_info != existing_info
+                ):
+                    raise RuntimeError(
+                        f"Info created by the info_spec {self.info_spec} is not equal to "
+                        f"info existing at '{self.path}' "
+                        "while `on_info_exists` is set to 'expect_same'"
+                    )
+                if existing_info != new_info:
+                    self.info_spec.write_info(self.path)
+            self.info_spec.reference_path = self.path
+            _info_cache[_info_hash_key(self.path)] = new_info
+            _ts_cache.clear()
```

Comment:
Wouldn't it be sufficient to just delete `self.path` from `_ts_cache`?

Reply:
Yeah - cachetools didn't support a clean way to do it which is why I'm just deleting the whole thing, but for bunk >> punk situation that's necessary. I'll implement it correctly before I push again.